feat: addressables

ShadauxCat opened the pull request:
MTT-2576
MTT-2578
MTT-2579
MTT-2580
MTT-2581
MTT-2582
MTT-2585

## Changelog

### com.unity.netcode.gameobjects
- Added: The SDK now supports loading assets from addressables
- Added: The SDK now supports modifying the NetworkPrefab list at runtime, so long as it's done before calling `StartServer()`/`StartClient()`/`StartHost()`

## Testing and Documentation

* Includes integration tests.
* Includes documentation for previously-undocumented public API entry points.

TwoTenPvP committed:
feat: Added addressables UI

ShadauxCat committed:
Addressables Runtime Support

ShadauxCat committed:
Merge branch 'addressables-ui' into feat/addressables

# Conflicts:
#	com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefab.cs
#	com.unity.netcode.gameobjects/Runtime/com.unity.netcode.runtime.asmdef

ShadauxCat committed:
Standards + changelog

ShadauxCat committed:
Addressables as a dependency

ShadauxCat committed:
Fixed missing preprocessor guards, removed hard dependency.

WhippetsAintDogs commented:
Hi @ShadauxCat, I have a little question for you. It seems like all network addressables are pulled into memory upon client / server / host start. Wouldn't it be better if they were to be loaded only when we need them (on spawned) ? I've implemented addressables support in our project using custom NetworkPrefabHandlers registered using the GlobalObjectIdHash value as key instead of a direct reference to the addressable prefab's NetworkObject, and we load the addressable in memory only upon spawning on the client side (in the handler's Instantiate method). What are your thoughts on the matter ?

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefab.cs:
```diff
@@ -41,5 +46,115 @@ internal class NetworkPrefab
+            }
+        }
+
+#if NETCODE_USE_ADDRESSABLES
```
This could be removed completely?

ShadauxCat replied:
uh... yeah. Yep. Yes. 👍 ✔️

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -491,6 +546,329 @@ private void OnValidate()
+                    if (NetworkConfig.NetworkTransport.StartServer())
+                    {
+                        State = NetworkManagerState.Ready;
+                        InvokeOnReadyCallback();
```
I am confused as to what this callback is for?
For the server-host it is invoked after the Transport is started, but a few lines down OnServerStarted is invoked.  So, for server-host it seems like it signifies that the transport has been started?
For the client it is invoked only if scene management is disabled and happens just before the InvokeOnClientConnectedCallback is invoked too... so if scene management is enabled then InvokeOnReadyCallback never is invoked for the client (is this intentional and why?).

This action seems like it is close to the same thing as OnServerStarted and InvokeOnClientConnectedCallback?

Perhaps some additional comments as to what InvokeOnReadyCallback is supposed to signify would help?

ShadauxCat replied:
We talked about this a little the other day, but just to put it in writing...

When addressables are in use, anything that needs them can't be used or spawned until after they've been loaded. Additionally, we've had issues in the past in general with people trying to do things before all the startup had been completed and it not working. This callback is one of two ways the user can know when the SDK is fully ready to use (connected, all addressable assets loaded, and for clients, connection has been approved by the server). The other way is for them to poll the State value in an Update event and watch for it to be Ready. Either way, it gives the user something they've been missing - an easy and reliable way to know when the SDK is ready.

With addressables, assets are loaded before the transport connection starts, which means that there's even more things that can't be done immediately after the return from `StartClient()`/`StartHost()`/`StartServer()`. That increases the need for the user to have access to this kind of information - that's why I chose now to add the state machine and callbacks, because the SDK with addressables would be very frustrating and difficult to use without this.

You are right that this tends to be invoked at the same time as `OnServerStarted` and `OnClientConnected`, but it's generic and the naming is clearer, to me. `OnClientConnected` is overloaded - sometimes it means "I successfully connected", other times it means "someone connected to me", making it more confusing to implement because you have to treat it differently between client and server, and you also have to know that Host will be treated like a server and will get `OnServerStarted` instead of `OnClientConnected`.

It's true the user could use those two callbacks to detect the same event, but I'm a fan of things that aren't confusing, and this callback can also be reused in both client and server code, while the others can't. If I could, I'd actually remove the other two, but since I can't do that in 1.0, I can at least add a clearer and easier-to-use alternative.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -11,11 +11,50 @@
 
 namespace Unity.Netcode
 {
+    public enum NetworkManagerState
```
❤️ 
I almost did this in a previous PR but then restrained myself.
Migrating towards a state driven NetworkManager is always a good thing in my book!

ShadauxCat committed:
Maybe fixes stand-alone test running.

ShadauxCat committed:
An attempt at ensuring assets are properly built for CI use. Who knows if this will work or not...

ShadauxCat committed:
standards

ShadauxCat committed:
- Removed unnecessary #if block
- Removed code that builds addressables on startup and on change

ShadauxCat committed:
- Improved resource management for addressables, should fix currently failing tests
- More tests
- Ensure all asset load failures call the OnStartupFailedCallback

ShadauxCat committed:
standards

ShadauxCat committed:
Nullreference fix when no config is present during shutdown.

ShadauxCat committed:
This fixes an unrelated test but I need to make sure the test failure isn't caused by addressables code.

ShadauxCat committed:
Merge branch 'develop' into feat/addressables

# Conflicts:
#	com.unity.netcode.gameobjects/CHANGELOG.md
#	com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs
#	com.unity.netcode.gameobjects/Runtime/com.unity.netcode.runtime.asmdef
#	com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTestHelpers.cs
#	testproject/Assets/Tests/Runtime/testproject.runtimetests.asmdef
#	testproject/Packages/manifest.json
#	testproject/ProjectSettings/EditorBuildSettings.asset

ShadauxCat committed:
Not sure how this managed to not get in the last commit.

ShadauxCat committed:
Standards

ShadauxCat committed:
Fix transition to Ready state when scene management is enabled.

ShadauxCat committed:
Merge branch 'develop' into feat/addressables

ShadauxCat committed:
Fix compile error when addressables isn't present.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefab.cs:
```diff
@@ -41,5 +52,112 @@ internal class NetworkPrefab
+    /// Class that represents a NetworkPrefab
+    /// </summary>
+    [Serializable]
+    internal class NetworkAddressable
```
Could this class be migrated into its own file?

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -349,6 +388,23 @@ public IReadOnlyList<ulong> ConnectedClientsIds
 
         internal void InvokeOnClientConnectedCallback(ulong clientId) => OnClientConnectedCallback?.Invoke(clientId);
 
+        public event Action OnReadyCallback = null;
```
It might be a good idea to add XML Doc summary for public facing API documentation reference purposes.
_(apply this to all public facing additions)_

NoelStephensUnity commented on testproject/Assets/Scripts/AddressablesBuilder.cs:
```diff
@@ -0,0 +1,19 @@
+#if UNITY_EDITOR
```
This needs to be wrapped in #if NETCODE_USE_ADDRESSABLES otherwise it won't compile if you remove the addressables package.

NoelStephensUnity commented on testproject/Assets/Tests/Runtime/AddressablesTests.cs:
```diff
@@ -0,0 +1,310 @@
+using System.Collections;
```
This needs to be wrapped in #if NETCODE_USE_ADDRESSABLES or it will not compile if the addessables package is not included in the project.

ShadauxCat replied:
Is there a reason we'd remove addressables from testproject?

NoelStephensUnity replied:
Well, this sort of brings up a broader topic about how we can test with and without addressables to make sure no future changes break the SDK when compiled without the addressables package included.  Since it is now part of our primary testproject, all of our tests will be run against NETCODE_USE_ADDRESSABLES defined builds.  We will need to most likely add a new undefined NETCODE_USE_ADDRESSABLES build in the nightly builds in order to run our tests against that.
So, we will need to be able to compile our test project when the addressable package is not included in order to be able to do this.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkConfig.cs:
```diff
@@ -30,13 +34,30 @@ public class NetworkConfig
+        /// The default player prefab addressable
+        /// </summary>
+        [Tooltip("When set, NetworkManager will automatically create and spawn the assigned player prefab. This can be overridden by adding it to the NetworkPrefabs list and selecting override.")]
+        public AssetReferenceGameObject PlayerAddressable;
```
Note to myself... I forgot to handle this. I need to add code to load this and set PlayerPrefab, and a test to make sure it's working.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -491,6 +547,369 @@ private void OnValidate()
         }
 #endif
 
+        public bool ResolveNetworkPrefabs()
```
Note to myself, this should be private.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -515,6 +934,9 @@ private void Initialize(bool server)
             MessagingSystem.Hook(new ProfilingHooks());
 #endif
 
+            // Uncomment to log every message that's sent and received
```
Personal opinion is that I would prefer not to have commented-out code that needs to be uncommented in order to have a message debugging tool...and if we were going to have something like that then we might as well make this minor modification? 
```
#if DEVELOPMENT_BUILD || UNITY_EDITOR
        internal INetworkHooks DebugLogHandler;
#endif
        private void Initialize(bool server)
        {
            // Don't allow the user to start a network session if the NetworkManager is
            // still parented under another GameObject
            if (NetworkManagerCheckForParent(true))
            {
                return;
            }

            if (NetworkLog.CurrentLogLevel <= LogLevel.Developer)
            {
                NetworkLog.LogInfo(nameof(Initialize));
            }

            this.RegisterNetworkUpdate(NetworkUpdateStage.EarlyUpdate);
            this.RegisterNetworkUpdate(NetworkUpdateStage.PostLateUpdate);

            MessagingSystem = new MessagingSystem(new NetworkManagerMessageSender(this), this);

            MessagingSystem.Hook(new NetworkManagerHooks(this));
#if DEVELOPMENT_BUILD || UNITY_EDITOR
            MessagingSystem.Hook(new ProfilingHooks());
            if (DebugLogHandler != null)
            {
                MessagingSystem.Hook(DebugLogHandler);
            }            
#endif
```
It could be made public as well...but only when in developer/editor mode.  Otherwise, it will all be internal anyways so really only SDK developers would use this capability and we can inject our own hooks into the messaging system directly...so maybe we just don't need to have this commented out code?

ShadauxCat replied:
The reason I don't have it working that way is that logging every single message sent and received is extremely, extremely spammy and makes the signal-to-noise ratio of the logs terrible. I left it here as a tool so I could stop re-implementing it and then deleting it every time I need to do that kind of debugging, but I don't think anyone would want to see that log output all the time, even in the editor.

NoelStephensUnity replied:
Right... so the above code will only add the hook if the DebugLogHandler is set...then that way...you can still add an instance of the DebugLogHooks when you want to see those messages... but then any other SDK developer could add that or make their own version...and one wouldn't need to recompile if you wanted to switch out DebugLogHandlers (i.e. command line parameter... UI drop down menu before starting server/host/client...etc) between sessions.

This way you keep the same functionality while also providing additional flexibility and ease of use.
When it is not set (default) then there are no hooks registered.

ShadauxCat replied:
That's fair, sorry, I misunderstood what you were suggesting. :) I'll do that.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkPrefabHandlerTests.cs:
```diff
@@ -76,6 +76,8 @@ public void NetworkConfigInvalidNetworkPrefabTest()
                 exceptionOccurred = true;
             }
 
+
```
You could delete these two lines and be -1 file modified.  
:)

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -588,216 +1010,32 @@ private void Initialize(bool server)
+        /// of startup success.
         /// </summary>
-        public bool StartServer()
+        public bool StartServer(Action onReady = null, Action<StartupFailureReason> onFailure = null)
```
Note to myself, should probably unregister these functions automatically at shutdown.

NoelStephensUnity reviewed (changes requested):
It all looks good!
There are just a few areas that need attention, and as long as we can compile test project without the addressables package to provide us with the ability/option to run our tests against that version of the SDK build I am good with this PR.

ShadauxCat committed:
- Addressed review feedback
- Fixed PlayerAddressable not being respected and added tests for it
- Split OnReadyCallback and OnStartupFailedCallback so that the ones provided as method parameters at startup are temporary but those assigned to the properties externally persist between startups.

NoelStephensUnity commented on testproject/Assets/Tests/Runtime/AddressablesTests.cs:
```diff
@@ -32,16 +33,60 @@ public enum AddMode
-    [TestFixture(LookupType.AssetByOperation, LoadMode.StartAsync, AddMode.AssetReference)]
-    [TestFixture(LookupType.AssetByOperation, LoadMode.WaitForFinish, AddMode.AssetReference)]
-    [TestFixture(LookupType.NetworkManagerByAddress, LoadMode.None, AddMode.String)]
+    public enum AddressableType
```
❤️‍🔥 
Like the use of TestFixture!

ShadauxCat replied:
Thanks!

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkAddressable.cs:
```diff
@@ -0,0 +1,101 @@
+#if NETCODE_USE_ADDRESSABLES
```
✔️

ShadauxCat committed:
Merge branch 'develop' into feat/addressables

ShadauxCat committed:
Replaced m_ShuttingDown with a more reliable state check, since state exists now.

ShadauxCat committed:
Merge remote-tracking branch 'origin/feat/addressables' into feat/addressables

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -379,7 +378,7 @@ public IReadOnlyList<ulong> ConnectedClientsIds
 
 
-        public bool ShutdownInProgress { get { return m_ShuttingDown; } }
+        public bool ShutdownInProgress { get { return State == NetworkManagerState.ShuttingDown; } }
```
💯

NoelStephensUnity committed:
Merge branch 'develop' into feat/addressables

0xFA11 commented on com.unity.netcode.gameobjects/Editor/com.unity.netcode.editor.asmdef:
```diff
@@ -13,6 +16,11 @@
+        {
+            "name": "com.unity.addressables",
+            "expression": "",
+            "define": "NETCODE_USE_ADDRESSABLES"
```
```suggestion
            "define": "COM_UNITY_ADDRESSABLES"
```

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/com.unity.netcode.runtime.asmdef:
```diff
@@ -28,6 +30,11 @@
+        {
+            "name": "com.unity.addressables",
+            "expression": "",
+            "define": "NETCODE_USE_ADDRESSABLES"
```
```suggestion
            "define": "COM_UNITY_ADDRESSABLES"
```

0xFA11 committed:
Merge branch 'develop' into feat/addressables

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -485,6 +586,399 @@ private void OnValidate()
+        }
+
+#if NETCODE_USE_ADDRESSABLES
+        public void AddNetworkPrefab(AssetReferenceGameObject addressableAsset)
```
`AddNetworkPrefab` -> `AddNetworkAddressable`?

0xFA11 replied:
also `xmldoc` on public APIs please?

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -485,6 +586,399 @@ private void OnValidate()
+
+            NetworkConfig.NetworkAddressables.Add(new NetworkAddressable { Addressable = addressableAsset });
+        }
+        public void AddNetworkPrefab(string address)
```
I think we should not have this method at all because it just wraps the ctor and it's confusing.

ShadauxCat replied:
My thought process here was that, if the SDK is offering to handle all the loading and management of the asset, that it might be simpler for the user to only keep the string around on their end. Yes, it just wraps the constructor, but it gives the user the ability to not have to worry about implementation details of addressables at all and just mark their things as addressables and work with only strings.

I'm not super attached to this function existing, but just wanted to mention my thought process and see what you think before I get rid of it.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -485,6 +586,399 @@ private void OnValidate()
+            return null;
+        }
+
+        public GameObject GetGameObjectForAddress(string address)
```
same, we probably should not have this method (just use the method above directly instead)

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -1232,11 +1551,18 @@ public void NetworkUpdate(NetworkUpdateStage updateStage)
 
+            if (State == NetworkManagerState.LoadingAssets)
+            {
+                if (!ResolveNetworkPrefabs())
```
I remember that we wanted to expose an async API to allow user customization. Something that users could chain their own async functions to set theirselves out before NetworkManager connects. As far as I can tell, this PR lacks that atm. Also, I think we could basically have an async API at this level and chain all async tasks/jobs down the line behind this function call, then just basically do `.IsValid()` & `.IsCompleted` here. That'd make the whole flow async without nesting "poll async state and return bool every earlyupdate".

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkAddressable.cs:
```diff
@@ -0,0 +1,101 @@
+            {
+                if (m_Prefab == null)
+                {
+                    throw new Exception("Not yet loaded.");
```
Can we make this message more clear?  Like "Addressable ____ not yet loaded."

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkAddressable.cs:
```diff
@@ -0,0 +1,101 @@
+                if (m_AsyncOperationHandle.Status == AsyncOperationStatus.Succeeded)
+                {
+                    var result = m_AsyncOperationHandle.Result;
+                    NetworkPrefab.VerifyValidPrefab(result);
```
It seems odd to me to have a "verify" function that doesn't return a value.  I see that it throws an exception which would bubble up to the caller of ResolveAsync.  I also see there's another exception thrown on line 84.  Is this what we want?  Seems a little hard to reason about.

ShadauxCat replied:
What would you suggest naming the function? CheckValidPrefab maybe?

ShadauxCat replied:
CheckIfPrefabIsValid?

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkAddressable.cs:
```diff
@@ -0,0 +1,101 @@
+                    throw new Exception($"Could not load addressable object: {Addressable.AssetGUID}");
+                }
+
+                return false;
```
So AsyncOperationStatus can be Failed, Succeeded of None.  Is None a legit state we can fall into?  Is None a 'false-y' condition but Failed is an exception-throwing condition?  Or can line 82 be an 'else' for all non-'Succeeded' cases?

ShadauxCat replied:
None = still loading. We just have to keep waiting when it's None.

ShadauxCat replied:
(Hence the return false)

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkAddressable.cs:
```diff
@@ -0,0 +1,101 @@
+        /// </summary>
+        public AssetReferenceGameObject OverridingTargetAddressable;
+
+        public bool ResolveAsync()
```
It looks like this function gets called by way of ResolveNetworkPrefabs()
both during StartHost/StartServer (once, obviously), then repeatedly in OnNetworkEarlyUpdate. 

I spent a considerable amount of time tracing all this out; at a minimum can you explain in comments the behavior of this function?  (e.g., this throws a "game over" exception if ____, returns true if ___, and returns false if ____ and should be called until it returns true)?

NoelStephensUnity replied:
_(xml doc public facing API)_

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -485,6 +586,399 @@ private void OnValidate()
 #endif
 
+        private bool ResolveNetworkPrefabs()
+        {
```
This function is pretty large - 300+ lines of code.  Can it be broken into smaller pieces?

ShadauxCat replied:
The vast majority of this code is moved code, not code I wrote... I'm always hesitant to try to squeeze these kinds of refactors of other code into unrelated changes just because I moved it from one place to another. This all used to be part of Initialize and had to be delayed until assets are loaded.

NoelStephensUnity replied:
My initial reaction was similar to Matt's until I realized the majority of that came from the Initialize method.  How I finally "convinced myself" that this is "ok" for now was that there would be no real functional point in trying to break it up into several functions inside of NetworkManager since this is all NetworkPrefab initialization related code that really should live somewhere outside of NetworkManager but still invoked by NetworkManager.
Now that it is all in its own function it will be easier to create a small task after this PR to migrate the code into a "NetworkPrefab" relative location...perhaps even beak it up into smaller chunks during or after that PR. 
_(like 1-2 minor PRs)_

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -485,6 +586,399 @@ private void OnValidate()
+                }
+                else
+                {
+                    if (NetworkConfig.NetworkTransport.StartServer())
```
I'm not crazy about having the starting of the transport is embedded in the ResolveNetworkPrefabs function

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -11,11 +11,50 @@
 
 namespace Unity.Netcode
 {
+    public enum NetworkManagerState
```
Feels like a natural thing to be in its own PR

ShadauxCat replied:
Had this discussion with Fatih. Could have been its own PR if I'd predicted the need for it and done it first... But addressables can't function without it, and it also has a dependency on Addressables unless I remove the LoadingAssets state and reimplement all the places the state gets changed. Splitting it at this point, to me, adds both time risk (unable to get this PR approved until that one's done and merged in) and code risk (the implementation of state on develop would look slightly different and that adds risk of partial bits of that ending up left in by accident after a merge)

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -345,6 +383,69 @@ public IReadOnlyList<ulong> ConnectedClientsIds
         internal void InvokeOnClientConnectedCallback(ulong clientId) => OnClientConnectedCallback?.Invoke(clientId);
 
+        /// <summary>
+        /// This callback is executed when the SDK is ready for use. Before this is executed, the SDK
```
This too I think belongs in its own PR / something worth a quick team sync on

ShadauxCat replied:
Same reponse as your other comment.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -485,6 +586,399 @@ private void OnValidate()
+            AddNetworkPrefab(new AssetReferenceGameObject(address));
+        }
+
+        public GameObject GetGameObjectForAddressable(AssetReferenceGameObject addressableAsset)
```
What would the use case for this public API be?  Do we need to expose this?

ShadauxCat replied:
Just trying to make it a little easier for users to use the API. I find the addressables API a little clunky and difficult to understand so I thought having something like this might make it easier - since we're doing all the rest of the addressables API under the hood, just felt to me like it was worthwhile to let the user just not think about the addressables API at all if they don't want to. Not opposed to removing it if you think that's superfluous.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -821,32 +1135,36 @@ public bool StartServer()
+            State = NetworkManagerState.LoadingAssets;
 
-            return false;
+            return ResolveNetworkPrefabs();
```
Again, per my earlier comment, embedding the transport start calls in ResolveNetworkPrefabs makes it harder to reason about / maintain this.

ShadauxCat replied:
Have a new proposal that will help decouple these, but will probably be a separate PR that builds on this one.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -1083,17 +1377,42 @@ public void Shutdown(bool discardMessageQueue = false)
                 NetworkLog.LogInfo(nameof(ShutdownInternal));
             }
 
+#if NETCODE_USE_ADDRESSABLES
```
Prefer having a "ShutdownAddressables()" method vs. embedding this here

wackoisgod commented:
> > Hi @ShadauxCat, I have a little question for you. It seems like all network addressables are pulled into memory upon client / server / host start. Wouldn't it be better if they were to be loaded only when we need them (on spawned) ? I've implemented addressables support in our project using custom NetworkPrefabHandlers registered using the GlobalObjectIdHash value as key instead of a direct reference to the addressable prefab's NetworkObject, and we load the addressable in memory only upon spawning on the client side (in the handler's Instantiate method). What are your thoughts on the matter ?
> 
> We had discussed a lot of options like this when we approached this feature. Ultimately we decided for the current release we didn't have the time to properly implement and fully test things like this. It adds a lot of complexity to the spawning system since the spawn - and everything else directed at that object - would have to be delayed until the load is complete, and we're not able to use synchronous loading because, not only is it a potentially very bad idea for performance (i.e., delaying a frame until an asset is downloaded from a remote source when the key doesn't point to a local asset bundle), but that API isn't supported on all versions and platforms that the SDK supports (for example, it doesn't work on WebGL)... so we'd have to refactor huge portions of the SDK to be able to handle asynchronous asset loading without resulting in messages being dropped or reordered in an invalid way.
> 
> That's not to say we won't support this in a future version (though I also can't promise that we will), but for this version we don't have the resources to make it work correctly and reliably for all users and thoroughly test all the use cases.

So I am curious... as someone who is invested in Unity's solution here, what was or is the point of just throwing in a solution and not actually doing the work to find the right solution? To me it just seems like a complete miss when it comes to wanting to provide customers any sort of value when you know it's not a great solution or even a tested solution.

As for the solution recommended by @WhippetsAintDogs  is there any reason why that solution doesn't make sense? Clearly the dev has gotten something working with whats released today. 

Also the purposed solution is pretty much useless for mobile devs because we don't want everything in memory on a low memory device and essentially with the current code it doesn't look like we have many options to really customize our loading strategy without just doing it ourselves much like how @WhippetsAintDogs did.

WhippetsAintDogs commented:
>  As for the solution recommended by @WhippetsAintDogs is there any reason why that solution doesn't make sense? Clearly the dev has gotten something working with whats released today.

Hi @wackoisgod ! For our use-case, we're loading addressables on a per "sectors" / "levels" basis. So when something is spawned at runtime inside a given sector, it is already loaded on both the clients and the server upon the given sector entry (and unloaded upon its exit) using `Addressables.LoadAssetAsync`. So when the PrefabHandler resolve the spawning on the client-side, using `operationHandle.WaitForCompletion()` to make the instantiation synchronous (in order to comply with the handler's Instantiate method signature) has virtually no cost since the asset is already pulled in memory. However our gain is not having direct references to these prefabs in the NetworkManager that would be pulled on the game start, instead they are pulled per sector.

So in a sense, our solution works for that kind of use-case but not use-cases where you're really only pulling something in memory upon instantiation / spawning - the cost of making the addressable flow synchronous would be heavy (it's a `while (true)` on the main thread under the hood 😅).

wackoisgod commented:
>Have to let @lpmaurice speak to this, because it wasn't my call here. I agree with you and tried to push for it but that's not the decision that was made. I don't want to speak for Louis here so I'll hope he can come and chime in to give more detail on the reasoning for you.

Indeed we can hope so :) 

>As it is, we know that the solution doesn't support all use cases for addressables. We will have a more robust solution in a future release, but this solution at least offers the ability to support simple use cases such as patch downloads, DLC content, etc that can be obtained and loaded before the SDK starts.

I mean I would argue this doesn't even support the most common platform that bundles are used on which is mobile and its used in such a way that speaks to exactly to what you mentioned which is the unloading and loading of things on the fly async. I mean more often then not developers are using bundles as a means to track and manage at a fine grain level whats loaded on a device and if I really don't have that control because the SDK is not flexible enough to allow for it it just seems like I would not want to use it it. 

I have said this in a past life but I do think the SDK should not care about what you are loading other than saying hey load this thing and should be an ID of some sort that is really up to the game to figure out what it needs to do with that. The real issue here is the SDK is so coupled to the the element of the prefab etc... that you can't have a representation of the network object separated from a prefab that would allow the simulation to continue to replicate for data and things load when they load for the visuals and sync up on that underlying network data. 

To me it just seems this solution is checking some boxes (not your fault @ShadauxCat) but not really solving a much larger issue which is the fact the SDK is so coupled to prefabs and  that essentially it forces game developers to adhear to whatever asset loading and data model it prescribes which just doesn't seem like its job.

ShadauxCat commented:
We've rethought our approach to addressables. This PR is being closed. The new approach is in #1882.
